src/main/webapp/WEB-INF/tags/viewedProducts.tag

```diff
@@ -0,0 +1,24 @@
+<%@ tag trimDirectiveWhitespaces="true" %>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+<style><%@ include file="/styles/main.css"%></style>
+
+<c:if test="${viewedProducts != null}">
+<jsp:useBean id="viewedProducts" type="com.es.phoneshop.model.entity.ProductHistory" scope="session"/>
```

Comment:
in filter you saved data to the request
then in service to session
and here we have scope="session"
something goes wrong within the implementation of recently viewed, please refactor